analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantArgument.cs

```diff
@@ -34,34 +36,26 @@ public sealed class RedundantArgument : SonarDiagnosticAnalyzer
             context.RegisterNodeAction(
                 c =>
                 {
-                    // Can't use optional arguments in expression trees (CS0584), so skip those
-                    if (c.IsInExpressionTree())
+                    if (!c.IsInExpressionTree() // Can't use optional arguments in expression trees (CS0584), so skip those
+                        && ArgumentList(c) is { Arguments.Count: > 0 } argumentList
+                        && new CSharpMethodParameterLookup(argumentList, c.SemanticModel) is { MethodSymbol: { } } methodParameterLookup)
                     {
-                        return;
-                    }
-
-                    var argumentList = c.Node is InvocationExpressionSyntax invocationExpression
-                        ? invocationExpression.ArgumentList
-                        : ((ImplicitObjectCreationExpressionSyntaxWrapper)c.Node).ArgumentList;
-                    var methodParameterLookup = new CSharpMethodParameterLookup(argumentList, c.SemanticModel);
-
-                    if (methodParameterLookup.MethodSymbol != null)
-                    {
-                        foreach (var argumentMapping in methodParameterLookup.GetAllArgumentParameterMappings().Reverse().Where(x => x.Symbol.HasExplicitDefaultValue))
+                        foreach (var argumentMapping in methodParameterLookup.GetAllArgumentParameterMappings().Reverse().Where(x => ArgumentHasDefaultValue(x, c.SemanticModel)))
                         {
-                            var hasDefaultValue = ArgumentHasDefaultValue(argumentMapping, c.SemanticModel);
-                            if (argumentMapping.Node.NameColon == null && !hasDefaultValue)
-                            {
-                                return;
-                            }
-                            else if (hasDefaultValue)
-                            {
-                                c.ReportIssue(Diagnostic.Create(Rule, argumentMapping.Node.GetLocation(), argumentMapping.Symbol.Name));
-                            }
+                            c.ReportIssue(Diagnostic.Create(Rule, argumentMapping.Node.GetLocation(), argumentMapping.Symbol.Name));
                         }
                     }
                 },
-                SyntaxKind.InvocationExpression, SyntaxKindEx.ImplicitObjectCreationExpression);
+                SyntaxKind.InvocationExpression, SyntaxKind.ObjectCreationExpression, SyntaxKindEx.ImplicitObjectCreationExpression);
+
+        private static ArgumentListSyntax ArgumentList(SonarSyntaxNodeReportingContext c) =>
```

Comment:
Why are we passing the entire context?
I think this is better:
```cs
private static ArgumentListSyntax ArgumentList(SyntaxNode node) =>
    node switch
    {
        InvocationExpressionSyntax invocationExpression => invocationExpression.ArgumentList,
        ObjectCreationExpressionSyntax objectCreationExpression => objectCreationExpression.ArgumentList,
        { RawKind: (int)SyntaxKindEx.ImplicitObjectCreationExpression } => ((ImplicitObjectCreationExpressionSyntaxWrapper)node).ArgumentList,
        _ => null,
    };
```

Reply:
Yes. That is definitely better. I will use the extension method of #8238. Therefore the PR is blocked at the moment.

https://github.com/SonarSource/sonar-dotnet/pull/8238/files#diff-02490849d0e39e1ae46c327f1ecac0cdc15fe8b032218bc918a0727287c3f199R336